File: backend/src/test/java/org/luxons/sevenwonders/game/data/serializers/ResourceTypesSerializerTest.java
package org.luxons.sevenwonders.game.data.serializers;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import org.junit.Before;
import org.junit.Test;
import org.luxons.sevenwonders.game.resources.ResourceType;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class ResourceTypesSerializerTest {

    private Gson gson;

    @Before
    public void setUp() {
        gson = new GsonBuilder().registerTypeAdapter(createListTypeToken(), new ResourceTypesSerializer()).create();
    }

    private static Type createListTypeToken() {
        return new TypeToken<List<ResourceType>>() {
        }.getType();
    }

    @Test
    public void serialize_null() {
        assertEquals("null", gson.toJson(null, createListTypeToken()));
    }

    @Test
    public void serialize_emptyList() {
        List<ResourceType> types = new ArrayList<>();
        assertEquals("\"\"", gson.toJson(types, createListTypeToken()));
    }

    @Test
    public void serialize_singleType() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        assertEquals("\"W\"", gson.toJson(types, createListTypeToken()));
    }

    @Test
    public void serialize_multipleTimesSameType() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        types.add(ResourceType.WOOD);
        types.add(ResourceType.WOOD);
        assertEquals("\"WWW\"", gson.toJson(types, createListTypeToken()));
    }

    @Test
    public void serialize_mixedTypes() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        types.add(ResourceType.CLAY);
        types.add(ResourceType.STONE);
        assertEquals("\"WCS\"", gson.toJson(types, createListTypeToken()));
    }

    @Test
    public void deserialize_null() {
        assertNull(gson.fromJson("null", createListTypeToken()));
    }

    @Test
    public void deserialize_emptyList() {
        List<ResourceType> types = new ArrayList<>();
        assertEquals(types, gson.fromJson("\"\"", createListTypeToken()));
    }

    @Test
    public void deserialize_singleType() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        assertEquals(types, gson.fromJson("\"W\"", createListTypeToken()));
    }

    @Test
    public void deserialize_multipleTimesSameType() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        types.add(ResourceType.WOOD);
        types.add(ResourceType.WOOD);
        assertEquals(types, gson.fromJson("\"WWW\"", createListTypeToken()));
    }

    @Test
    public void deserialize_mixedTypes() {
        List<ResourceType> types = new ArrayList<>();
        types.add(ResourceType.WOOD);
        types.add(ResourceType.CLAY);
        types.add(ResourceType.STONE);
        assertEquals(types, gson.fromJson("\"WCS\"", createListTypeToken()));
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/data/serializers/ResourceTypesSerializerTest.kt
package org.luxons.sevenwonders.game.data.serializers

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.luxons.sevenwonders.game.resources.ResourceType
import java.lang.reflect.Type
import java.util.ArrayList

class ResourceTypesSerializerTest {

    private lateinit var gson: Gson
    
    @Before
    fun setUp() {
        gson = GsonBuilder().registerTypeAdapter(createListTypeToken(), ResourceTypesSerializer()).create()
    }

    private fun createListTypeToken(): Type {
        return object : TypeToken<List<ResourceType>>() {}.type
    }

    @Test
    fun serialize_null() {
        assertEquals("null", gson.toJson(null, createListTypeToken()))
    }

    @Test
    fun serialize_emptyList() {
        val types = ArrayList<ResourceType>()
        assertEquals("\"\"", gson.toJson(types, createListTypeToken()))
    }

    @Test
    fun serialize_singleType() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        assertEquals("\"W\"", gson.toJson(types, createListTypeToken()))
    }

    @Test
    fun serialize_multipleTimesSameType() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        types.add(ResourceType.WOOD)
        types.add(ResourceType.WOOD)
        assertEquals("\"WWW\"", gson.toJson(types, createListTypeToken()))
    }

    @Test
    fun serialize_mixedTypes() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        types.add(ResourceType.CLAY)
        types.add(ResourceType.STONE)
        assertEquals("\"WCS\"", gson.toJson(types, createListTypeToken()))
    }

    @Test
    fun deserialize_null() {
        assertNull(gson.fromJson("null", createListTypeToken()))
    }

    @Test
    fun deserialize_emptyList() {
        val types = ArrayList<ResourceType>()
        assertEquals(types, gson.fromJson("\"\"", createListTypeToken()))
    }

    @Test
    fun deserialize_singleType() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        assertEquals(types, gson.fromJson("\"W\"", createListTypeToken()))
    }

    @Test
    fun deserialize_multipleTimesSameType() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        types.add(ResourceType.WOOD)
        types.add(ResourceType.WOOD)
        assertEquals(types, gson.fromJson("\"WWW\"", createListTypeToken()))
    }

    @Test
    fun deserialize_mixedTypes() {
        val types = ArrayList<ResourceType>()
        types.add(ResourceType.WOOD)
        types.add(ResourceType.CLAY)
        types.add(ResourceType.STONE)
        assertEquals(types, gson.fromJson("\"WCS\"", createListTypeToken()))
    }
}
